### ISSUE TITLE
[BUG]: MPNN.encoding() doesn't work

### ISSUE THREAD

**Issue by shihchengli** [chemprop/chemprop#754]
**Describe the bug**
`MPNN.encoding()` is designed to obtain the final hidden representations for the input molecules. However, when calling this function, an error will occur.

**Example(s)**
```
import pandas as pd
import numpy as np
import torch
from lightning import pytorch as pl
from chemprop import data, featurizers, models

# Load model
checkpoint_path = '../tests/data/example_model_v2_regression_mol.ckpt' # path to the checkpoint file. 
mpnn = models.MPNN.load_from_checkpoint(checkpoint_path)

# Load data
test_path = '../tests/data/smis.csv'
smiles_column = 'smiles'
df_test = pd.read_csv(test_path)
smis = df_test[smiles_column]
test_data = [data.MoleculeDatapoint.from_smi(smi) for smi in smis]
featurizer = featurizers.SimpleMoleculeMolGraphFeaturizer()
test_dset = data.MoleculeDataset(test_data, featurizer=featurizer)
test_loader = data.MolGraphDataLoader(test_dset, shuffle=False)

# Get encoding
all_fingerprints = []
with torch.no_grad():
    for batch in test_loader: 
        bmg, V_d, X_d, *_ = batch  # Update these based on your batch structure
        fingerprints = mpnn.encoding(bmg, V_d, X_d)
all_fingerprints.append(fingerprints)
```

**Error Stack Trace**
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[11], line 20
     18     for batch in test_loader: 
     19         bmg, V_d, X_d, *_ = batch  # Update these based on your batch structure
---> 20         fingerprints = mpnn.encoding(bmg, V_d, X_d)
     21 all_fingerprints.append(fingerprints)

File ~/Documents/MIT/chemprop/chemprop/models/model.py:131, in MPNN.encoding(self, bmg, V_d, X_d)
    127 def encoding(
    128     self, bmg: BatchMolGraph, V_d: Tensor | None = None, X_d: Tensor | None = None
    129 ) -> Tensor:
    130     """the final hidden representations for the input molecules"""
--> 131     return self.predictor[:-1](self.fingerprint(bmg, V_d, X_d))

TypeError: 'RegressionFFN' object is not subscriptable
```

**Additional context**
An instance of the class `RegressionFFN` does not support slicing operations. I think the code `self.predictor[:-1]` on [line 131](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/models/model.py#L131) in model.py should be modified to `self.predictor.ffn[:-1]`. This is because `ffn` is an instance of the class `MLP`, and `MLP` inherits from `nn.Sequential`, which allows the creation of a new instance of the same class with the sliced modules. Upon testing, however, another error appears:`TypeError: MLP.__init__() missing 1 required positional argument: 'output_dim'`. 

This error is due to when I use slicing on a module that inherits from `nn.Sequential`, it tries to create a new instance of the same class with the sliced modules. This works seamlessly for `nn.Sequential` because it doesn't require any additional arguments besides the modules themselves. However, in the `MLP` class, the constructor expects specific arguments (`input_dim`, `output_dim`, etc.), and slicing does not automatically pass these arguments to the new instance, leading to a TypeError. I have two ideas to solve it:
(1) `Override the __getitem__ Method:`
Override the `__getitem__` method in the `MLP` class to handle slicing properly.
```
def __getitem__(self, idx):
    if isinstance(idx, slice):
        # Handle slicing: create a new nn.Sequential with the sliced modules
        sliced_modules = OrderedDict(list(self._modules.items())[idx])
        return nn.Sequential(sliced_modules)
    else:
        # Handle single index access
        return self._get_item_by_idx(self._modules.values(), idx)
```
(2) `Access the underlying sequential directly:`
If the `MLP` class is composed of layers stored in an `nn.Sequential` attribute, access this attribute directly for slicing. This way, we can avoid the problem altogether but lose the specific functionalities of the `MLP` class for the sliced part. So, we would need to use `self.predictor.ffn.layers[:-1]` in `MPNN.encoding()`...
```
class MLP(nn.Module):
    def __init__(self, input_dim, output_dim, hidden_dim=300, n_layers=1, dropout=0.0, activation="relu"):
        super(MLP, self).__init__()
        self.input_dim = input_dim
        self.output_dim = output_dim

        # Create the layers list
        layers = [nn.Dropout(dropout), nn.Linear(input_dim, hidden_dim), get_activation_function(activation)]
        for _ in range(n_layers - 1):
            layers += [nn.Dropout(dropout), nn.Linear(hidden_dim, hidden_dim), get_activation_function(activation)]
        layers += [nn.Linear(hidden_dim, output_dim)]

        # Store the layers in an nn.Sequential
        self.layers = nn.Sequential(*layers)

    def forward(self, x):
        return self.layers(x)
```

**Comment by davidegraff** [comment]
IMO just delegate this method to the `Predictor`. That is:
```python
class MPNN:
    ...
    def encode(self, bmg):
        Z = self.encode(Z)
        return self.predictor.encoding(Z)
```
This passes the responsibility of calulating an "encoding" down to the individual predictor. To me, this makes sense because we define an encoding as the "final hidden representation of the FFN" meaning that it's implicitly a function defined by the predictor and **not** at the level of the MPNN

**Comment by shihchengli** [comment]
I agree that we can delegate this task to individual predictors. However, within the predictor, we still need to exclude the last layer from the MLP to obtain the final hidden representation. Do you have any thoughts on how to address the issue related to `nn.Sequential` mentioned above?

**Comment by shihchengli** [comment]
Treating the MLP as a sequence of blocks sounds good; however, this approach prevents us from extracting the final hidden representation, which serves as the input to the last layer. By using blocks, we face difficulties in indexing it correctly because we consider three layers as a single block.

Defining a class method could solve the problem, but it would be inconsistent with the way we create the `message_passing` block, where we don't define a `build` function. I think we can simply define a `__getitem__` function in the `MLP` class.

**Comment by davidegraff** [comment]
- The final hidden representation in this new object model is just `MLP[:-1]`. Ex: given `input_dim=64`, `hidden_dim=128`, `output_dim=128`, and `n_layers=2` the previous model produces an FFN like so:

```
Sequential(
  (0): Linear(in_features=64, out_features=128, bias=True)
  (1): ReLU()
  (2): Dropout(p=0.1, inplace=False)
  (3): Linear(in_features=128, out_features=128, bias=True)
  (4): ReLU()
  (5): Dropout(p=0.1, inplace=False)
  (6): Linear(in_features=128, out_features=2, bias=True)
)
```
and the new model:
```
Sequential(
  (0): Sequential(
    (0): Linear(in_features=64, out_features=128, bias=True)
  )
  (1): Sequential(
    (0): ReLU()
    (1): Dropout(p=0.1, inplace=False)
    (2): Linear(in_features=128, out_features=128, bias=True)
  )
  (2): Sequential(
    (0): ReLU()
    (1): Dropout(p=0.1, inplace=False)
    (2): Linear(in_features=128, out_features=2, bias=True)
  )
)
```
The organization is the same in both so the final hidden representation is the same as well (`MLP[:4]` vs `MLP[:-1]`).

- The two classes are not comparable in the slightest. There's no need to define parallel initialization structures. One is convolutional operation defined on graphs and the other is syntactic sugar for `nn.Sequential` to make our lives easier.

**Comment by shihchengli** [comment]
I think we are using MLP[:-1] for both v1 and the current v2 as the encoding (i.e., `--fingerprint_type last_FFN` in v1), which I believe is `MLP[:6]` instead of `MLP[:4]`.

**Comment by davidegraff** [comment]
Then it just depends how you define the blocks. You can easily rearrange the code such that `blocks[-1]` gets you the same result. Though I'm not really sure why you would define "the final hidden" this way rather than using the raw linear outputs. That is, the current definition also applies an activation and a dropout before returning them, meaning that you restrict these "fingerprints" to be positive semi-definite and limiting their expressivity. It's not a real issue, it just doesn't make much sense to me from a client's perspective.

**Comment by shihchengli** [comment]
Actually, I also feel that returning the raw linear outputs makes more sense to me. The reason I mentioned `MLP[:6]` is because I think that v1 was implemented this way. If we have a consensus on using the `MLP[:4]`, I can work on a PR to address the MPNN.encoding() issue.

**Comment by davidegraff** [comment]
fixed by #760
